### PR TITLE
New implementation of the data abstraction layer

### PR DESCRIPTION
**Pull request by mrkn**
In this pull-request, I'd like to rewrite data abstraction layer.

### CODE REVIEW

**Comment by mrkn** [comment]
@kou @284km How do you think this approach?

**Comment by kou** [comment]
I'll take a look this in this week.

File: lib/charty/table_adapters.rb [hunk 1]
```diff
@@ -0,0 +1,23 @@
+module Charty
+  module TableAdapters
+    @adapters = {}
+
+    def self.register(name, adapter)
+      @adapters[name] = adapter
+    end
+
+    def self.lookup_adapter_maker(data)
```
Comment (284km) [inline lib/charty/table_adapters.rb]: I think method name "find" is more intuitive than "lookup_adapter_maker". How about this?

File: lib/charty/table_adapters/nmatrix_adapter.rb [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        return index if index
+        raise IndexError, "Invalid column name: #{name}"
+      end
+
+      private def generate_column_names(n_columns, columns)
+        columns ||= []
+        if columns.length >= n_columns
+          columns[0, n_columns]
+        else
+          columns + columns.length.upto(n_columns - 1).map {|i| "X#{i}" }
```
Comment (284km) [inline lib/charty/table_adapters/nmatrix_adapter.rb]: I think it is appropriate to fill in column names with "X0", "X1"... "Xx".

File: test/table/nmatrix_test.rb [hunk 3]
```diff
@@ -0,0 +1,116 @@
+require 'test_helper'
+require 'numo/narray'
```
Comment (284km) [inline test/table/nmatrix_test.rb]: chore: `require 'nmatrix'`

File: lib/charty/table_adapters/datasets_adapter.rb [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+      end
+
+      # @param [Integer] i  Row index
+      # @param [Symbol,String,Integer] j  Column index
+      def [](i, j)
+        fetch_all if @records.empty?
+        @records[i][j]
+      end
+
+      class GenericDatasetsAdapter < self
```
Comment (284km) [inline lib/charty/table_adapters/datasets_adapter.rb]: I think I don't understand the need for "GenericDatasetsAdapter class" at this point. In what cases is it effective?

**Comment by 284km** [comment]
@mrkn Thanks for the explanation in the code.
I asked a few questions but I agree with the direction of this approach.

File: lib/charty/table_adapters/datasets_adapter.rb (hunk 4 above)
Comment (mrkn) [reply lib/charty/table_adapters/datasets_adapter.rb]: I introduced this adapter for subclasses of Datasets which have their own Record class.
I've renamed it to RecordCollectionAdapter.

Comment (mrkn) [reply lib/charty/table_adapters/datasets_adapter.rb]: The reason why I introduce this class as a subclass of DatasetsAdapter is that I want to introduce adapters for MNIST, CIFAR, libsvm, etc. in the future.

File: lib/charty/table_adapters.rb (hunk 1 above)
Comment (mrkn) [reply lib/charty/table_adapters.rb]: Renamed it to `find_adapter_maker`.

**Comment by kou** [comment]
I realize that we need to improve `Datasets::Table` API.
I'm working on it. Please wait for a while.

**Comment by mrkn** [comment]
@kou Yes. I don’t consider this pull-request provides complete Table API. We need improve this new Table API. For example, we should allow accessing columns by String and Symbol whatever the classes of column name.

But I consider this is enough to start working on rewriting plotting layer.
So I want to merge this before improvement.

**Comment by mrkn** [comment]
@kou But if you consider this approach is not fine, I’ll withdraw this pull-request.

**Comment by kou** [comment]
OK. Basically, I like this approach but I want to simplify API. How about the following?

```diff
diff --git a/lib/charty/table.rb b/lib/charty/table.rb
index 5f2e8dc..9d587f8 100644
--- a/lib/charty/table.rb
+++ b/lib/charty/table.rb
@@ -15,8 +15,12 @@ module Charty
     extend Forwardable
 
     def initialize(data, **kwargs)
-      adapter_maker = TableAdapters.find_adapter_maker(data)
-      @adapter = adapter_maker.make(data, **kwargs)
+      adapter_class = TableAdapters.find_adapter_class(data)
+      if kwargs.empty?
+        @adapter = adapter_class.new(data)
+      else
+        @adapter = adapter_class.new(data, **kwargs)
+      end
     end
 
     attr_reader :adapter
diff --git a/lib/charty/table_adapters.rb b/lib/charty/table_adapters.rb
index 86d880e..464755b 100644
--- a/lib/charty/table_adapters.rb
+++ b/lib/charty/table_adapters.rb
@@ -2,13 +2,13 @@ module Charty
   module TableAdapters
     @adapters = {}
 
-    def self.register(name, adapter)
-      @adapters[name] = adapter
+    def self.register(name, adapter_class)
+      @adapters[name] = adapter_class
     end
 
-    def self.find_adapter_maker(data)
-      @adapters.each_value do |adapter|
-        return adapter if adapter.supported?(data)
+    def self.find_adapter_class(data)
+      @adapters.each_value do |adapter_class|
+        return adapter_class if adapter_class.supported?(data)
       end
       raise ArgumentError, "Unsupported data class: #{data.class}"
     end
diff --git a/lib/charty/table_adapters/active_record_adapter.rb b/lib/charty/table_adapters/active_record_adapter.rb
index e3f4d78..d6c67f7 100644
--- a/lib/charty/table_adapters/active_record_adapter.rb
+++ b/lib/charty/table_adapters/active_record_adapter.rb
@@ -3,10 +3,6 @@ module Charty
     class ActiveRecordAdapter
       include Enumerable
 
-      def self.make(data)
-        self.new(data)
-      end
-
       def self.supported?(data)
         defined?(ActiveRecord::Relation) && data.is_a?(ActiveRecord::Relation)
       end
diff --git a/lib/charty/table_adapters/daru_adapter.rb b/lib/charty/table_adapters/daru_adapter.rb
index 4ec8588..a1e8a17 100644
--- a/lib/charty/table_adapters/daru_adapter.rb
+++ b/lib/charty/table_adapters/daru_adapter.rb
@@ -1,13 +1,8 @@
 module Charty
   module TableAdapters
     class DaruAdapter
-      extend Forwardable
       include Enumerable
 
-      def self.make(data)
-        self.new(data)
-      end
-
       def self.supported?(data)
         defined?(Daru::DataFrame) && data.is_a?(Daru::DataFrame)
       end
diff --git a/lib/charty/table_adapters/datasets_adapter.rb b/lib/charty/table_adapters/datasets_adapter.rb
index 65cbed1..03d2c08 100644
--- a/lib/charty/table_adapters/datasets_adapter.rb
+++ b/lib/charty/table_adapters/datasets_adapter.rb
@@ -1,20 +1,12 @@
 module Charty
   module TableAdapters
     class DatasetsAdapter
-      extend Forwardable
       include Enumerable
 
-      def self.make(dataset)
-        case
-        when dataset.class.const_defined?(:Record)
-          RecordCollectionAdapter.new(dataset)
-        else
-          raise TypeError, "Unsupported dataset class: #{dataset.class}"
-        end
-      end
-
       def self.supported?(data)
-        defined?(Datasets::Dataset) && data.is_a?(Datasets::Dataset)
+        defined?(Datasets::Dataset) &&
+          data.is_a?(Datasets::Dataset) &&
+          data.class.const_defined?(:Record)
       end
 
       def initialize(dataset)
@@ -22,6 +14,10 @@ module Charty
         @records = []
       end
 
+      def columns
+        @dataset.class::Record.members
+      end
+
       def each(&block)
         return to_enum(__method__) unless block_given?
 
@@ -46,12 +42,6 @@ module Charty
         fetch_all if @records.empty?
         @records[i][j]
       end
-
-      class RecordCollectionAdapter < self
-        def columns
-          @dataset.class::Record.members
-        end
-      end
     end
 
     register(:datasets, DatasetsAdapter)
diff --git a/lib/charty/table_adapters/hash_adapter.rb b/lib/charty/table_adapters/hash_adapter.rb
index f953cc2..47ad6b5 100644
--- a/lib/charty/table_adapters/hash_adapter.rb
+++ b/lib/charty/table_adapters/hash_adapter.rb
@@ -7,10 +7,6 @@ module Charty
       extend Forwardable
       include Enumerable
 
-      def self.make(data)
-        new(data)
-      end
-
       def self.supported?(data)
         case data
         when []
diff --git a/lib/charty/table_adapters/narray_adapter.rb b/lib/charty/table_adapters/narray_adapter.rb
index 3101e3a..90d4bb3 100644
--- a/lib/charty/table_adapters/narray_adapter.rb
+++ b/lib/charty/table_adapters/narray_adapter.rb
@@ -1,10 +1,6 @@
 module Charty
   module TableAdapters
     class NArrayAdapter
-      def self.make(data)
-        self.new(data)
-      end
-
       def self.supported?(data)
         defined?(Numo::NArray) && data.is_a?(Numo::NArray) && data.ndim <= 2
       end
diff --git a/lib/charty/table_adapters/nmatrix_adapter.rb b/lib/charty/table_adapters/nmatrix_adapter.rb
index bf8afa7..2a34b8e 100644
--- a/lib/charty/table_adapters/nmatrix_adapter.rb
+++ b/lib/charty/table_adapters/nmatrix_adapter.rb
@@ -1,10 +1,6 @@
 module Charty
   module TableAdapters
     class NMatrixAdapter
-      def self.make(data)
-        self.new(data)
-      end
-
       def self.supported?(data)
         defined?(NMatrix) && data.is_a?(NMatrix) && data.shape.length <= 2
       end
diff --git a/test/table/red_datasets_test.rb b/test/table/red_datasets_test.rb
index 6632be3..401cfbf 100644
--- a/test/table/red_datasets_test.rb
+++ b/test/table/red_datasets_test.rb
@@ -4,7 +4,7 @@ require 'datasets'
 class TableRedDatasetsTest < Test::Unit::TestCase
   sub_test_case("Charty::Table.new") do
     test("CIFAR dataset is unsupported yet") do
-      assert_raise(TypeError) do
+      assert_raise(ArgumentError) do
         Charty::Table.new(Datasets::CIFAR.new)
       end
     end
diff --git a/test/table_adapters_test.rb b/test/table_adapters_test.rb
index ddf95bf..6542aa2 100644
--- a/test/table_adapters_test.rb
+++ b/test/table_adapters_test.rb
@@ -3,14 +3,14 @@ require 'daru'
 require 'numo/narray'
 
 class TableAdaptersTest < Test::Unit::TestCase
-  sub_test_case(".find_adapter_maker") do
+  sub_test_case(".find_adapter_class") do
     test("for a hash of arrays") do
       data = {
                "foo" => [1, 2, 3, 4],
                "bar" => [5, 6, 7, 8],
              }
       assert_equal(Charty::TableAdapters::HashAdapter,
-                   Charty::TableAdapters.find_adapter_maker(data))
+                   Charty::TableAdapters.find_adapter_class(data))
     end
 
     test("for an array of hashes") do
@@ -21,7 +21,7 @@ class TableAdaptersTest < Test::Unit::TestCase
                {"foo" => 4, "bar" => 8},
              ]
       assert_equal(Charty::TableAdapters::HashAdapter,
-                   Charty::TableAdapters.find_adapter_maker(data))
+                   Charty::TableAdapters.find_adapter_class(data))
     end
 
     test("for an array of arrays") do
@@ -30,7 +30,7 @@ class TableAdaptersTest < Test::Unit::TestCase
                [5, 6, 7, 8]
              ]
       assert_equal(Charty::TableAdapters::HashAdapter,
-                   Charty::TableAdapters.find_adapter_maker(data))
+                   Charty::TableAdapters.find_adapter_class(data))
     end
 
     test("for an array of Numo::NArray arrays") do
@@ -39,13 +39,13 @@ class TableAdaptersTest < Test::Unit::TestCase
                [5, 6, 7, 8]
              ]
       assert_equal(Charty::TableAdapters::HashAdapter,
-                   Charty::TableAdapters.find_adapter_maker(data))
+                   Charty::TableAdapters.find_adapter_class(data))
     end
 
     test("for an array of scalar values") do
       data = [1, 2, 3, 4]
       assert_equal(Charty::TableAdapters::HashAdapter,
-                   Charty::TableAdapters.find_adapter_maker(data))
+                   Charty::TableAdapters.find_adapter_class(data))
     end
 
     test("for a Daru::DataFrame") do
@@ -54,7 +54,7 @@ class TableAdaptersTest < Test::Unit::TestCase
         "bar" => [5, 6, 7, 8]
       )
       assert_equal(Charty::TableAdapters::DaruAdapter,
-                   Charty::TableAdapters.find_adapter_maker(data))
+                   Charty::TableAdapters.find_adapter_class(data))
     end
 
     test("for a Numo::NArray matrix") do
@@ -65,7 +65,7 @@ class TableAdaptersTest < Test::Unit::TestCase
                           [4, 8, 12],
                         ]
       assert_equal(Charty::TableAdapters::NArrayAdapter,
-                   Charty::TableAdapters.find_adapter_maker(data))
+                   Charty::TableAdapters.find_adapter_class(data))
     end
 
     test("for a NMatrix matrix") do
@@ -77,7 +77,7 @@ class TableAdaptersTest < Test::Unit::TestCase
                            4, 8, 12,
                          ])
       assert_equal(Charty::TableAdapters::NMatrixAdapter,
-                   Charty::TableAdapters.find_adapter_maker(data))
+                   Charty::TableAdapters.find_adapter_class(data))
     end
   end
 end
```

File: lib/charty/table.rb [hunk 5]
```diff
@@ -1,11 +1,41 @@
...
+
+    def initialize(data, **kwargs)
+      adapter_maker = TableAdapters.find_adapter_maker(data)
+      @adapter = adapter_maker.make(data, **kwargs)
     end
 
-    attr_reader :table
+    attr_reader :adapter
+
+    def_delegator :@adapter, :columns
```
Comment (kou) [inline lib/charty/table.rb]: How about `column_names`?

File: lib/charty/table.rb [hunk 6]
```diff
@@ -1,11 +1,41 @@
...
+
+    def arrays
```
Comment (kou) [inline lib/charty/table.rb]: How about `columns`?

File: lib/charty/table_adapters/active_record_adapter.rb [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+
+      def initialize(data)
+        @data = check_type(ActiveRecord::Relation, data, :data)
+        @columns = @data.column_names.freeze
+        @arrays = nil
+      end
+
+      attr_reader :columns
+
+      def column(i)
```
Comment (kou) [inline lib/charty/table_adapters/active_record_adapter.rb]: How about `column_data` or `column_values`?
I like `fetch_column_XXX` or something but it's long...

Is `#[](column_index)` enough?

**Comment by kou** [comment]
Ah, I think that `TableAdapters`'s `@adapters` should be `@adapter_classes`.

**Pull request merged**